### DISCUSSION TITLE
Add volume_step data parameter to media_player.volume_up / volume_down service calls

### DISCUSSION THREAD

**Discussion by glenrobertson** [home-assistant/architecture#891] [Entity Models] [answered]
Hi all!

I opened https://github.com/home-assistant/core/pull/90283 to add a volume_step parameter to the media_player.volume_up / media_player.volume_down service calls.
Currently there is a 10% value hardcoded across all media_player entities, so this would allow a caller to configure that, as follows:

<img src="https://user-images.githubusercontent.com/379485/236856677-159f47d5-49c6-4f17-a79c-dfce5d59543c.png" width="400" />

Any feedback or discussion appreciated. Thanks in advance!

**Comment by mvdwetering** [comment]
There are muiltiple integrations that communicate with devices that offer a "volume-step-up/down" command. Those commands are mostly that, just step up/down without a way to indicate stepsize. 

Allowing to pass in arbitrary values through the service call makes it impossible to use those native step commands. By providing arbitrary values basically all you can do is take current volume and add or subtract the step value. Which is also how some integrations implement the VOLUME_STEP feature. 
However that would not work if your media_player can _only_ support VOLUME_STEP and not VOLUME_SET. Which is possible to indicate with supported_features, but I did not check if such media_players currently exist in HA.

**I think the real question is: shouldn't integrations that indicate support for volume step implement a proper stepsize?**

Question then of course becomes what a proper stepsize is. Which in turn depends on what the intended use of the VOLUME_STEP feature is.

I don't know what the original background of the VOLUME_STEP feature is, but I would expect to be able to control the volume like I would with a TV remote control or the volume up/down buttons on my PC keyboard. These are typically small steps to make the sound a bit louder/less loud. 

Maybe lets look at some examples of devices that offer volume step up/down functionality (just picked some I have easy access to):

* Windows 11, range 0-100, stepsize 2 = 2%
* Sonos App, range 0-100, stepsize 2 or 4 (stepsize seems to change for higher volumes) = 2% or 4%
* Android phone volume for media playback, range is 0-25 it seems, stepsize = 1 = 4%
* Yamaha receiver, -80.5 to 16.5dB, stepsize 0.5dB = about 0.5%

Maybe the default stepsize should just be smaller than 10% instead of making it variable?

That would give more steps for all integrations using the default implementation (a very quick search seems to indicate ~70 integrations support VOLUME_STEP feature and about 50 integrations have an implementation for `(async)_volume_up`). 
Note that very small default steps might give issues with integrations where the achieveable stepssize on the device is smaller than the default stepsize chosen. But then again they can be fixed by providing a custom implementation for volume_up/down.

(on a sidenote, it looks like the media_player group also does the 10% steps and not call the volume up/down of its members, that might also need updating)

I am sorry I don't ahve a direct solution and it got a bit longer than expected when I started writing, but I kept finding new stuff. I hope this is of some help.

**Reply by glenrobertson** [reply]
Thank you for replying and providing the context. 
I think I agree that the integrations should provide a sensible step size. With this change, I figured that users may want to customize it easily without having to delve into the integration code
Like even if my receiver has a sensible step up or down, I may wish to half or double it to my liking.
Happy to close if you disagree or merge if you do agree.

Thanks again

**Reply by mvdwetering** [reply]
> Happy to close if you disagree or merge if you do agree.

I was just tyring to provide some additonal scenarios. I don't have the power to actually decide on stuff  :)

**Comment by glenrobertson** [comment]
Nobody has commented on this so I'm just gonna close. Can re open if there is interest.

**Comment by gjohansson-ST** [comment]
I'll re-open this as the topic is still valid to take a decision upon.

**Comment by glenrobertson** [comment]
Closing but I can't close the other PR I opened so someone else is gonna have to do that lol.

**Reply by gjohansson-ST** [reply]
Not sure why you're still trying to close this. It's still valid I believe

**Reply by emontnemery** [reply]
Yes, this is indeed still valid 👍

There's a problem with the proposal though; the reason behind the `media_player.volume_up` / `media_player.volume_down` services is that some media players don't allow setting an absolute volume. An example is an amplifier with a motorized volume knob which does not know its position which means it's only possible to request the volume to be increased or decreased by an unknown amount.

Hence, it would be better to add a `volume_step` parameter to `media_player.volume_set` instead, where `volume_step` would be a positive or negative float. This would be similar to `brightness_step` for `light.turn_on`.

**Reply by glenrobertson** [reply]
Ok thanks for addressing it. I had closed as it was stale and I personally don't like keeping PRs open for 6 months, but appreciate you finally getting to this.
I can reopen the PR and implement how you've suggested.

**Comment by emontnemery** [comment, marked as answer]
I discussed this with @balloob, @edenhaus and @MartinHjelmare. The conclusion is that the proposal is fine; the `volume_step` service call parameter should just be filtered in `service.yaml` for media players which don't support `media_player.volume_set`.

**Reply by gjohansson-ST** [reply]
As far as I know you can't do negative filtering e.g. only showing the field if entity does not have `media_player.volume_set`.
Or am I missing what you meant?

**Reply by andrewsayre** [reply]
I think he meant filter the parameter by support for `VOLUME_STEP`, something like this:

```yaml
volume_set:
  ...
  fields:
    ...
    volume_step:
      filter:
        supported_features:
          - media_player.MediaPlayerEntityFeature.VOLUME_STEP
      required: false
      selector:
        number:
          min: 0
          max: 1
          step: 0.01
```

**Reply by andrewsayre** [reply]
@emontnemery Why not also add `volume_step` as an optional parameter to the volume up/down services as well? Those services already require both `VOLUME_SET` and `VOLUME_STEP`. If a platform doesn't support a custom step size it was already using a custom implementation of those actions and (likely) not using the built in 10% default anyway. It can just ignore the field, or would it be better to raise a `ServiceValidationException`?

**Reply by emontnemery** [reply]
What I meant was that a custom step size can only be supported by media players which support setting an absolute volume, so something like this:
```yaml
volume_set:
  ...
  fields:
    ...
    volume_step:
      filter:
        supported_features:
          - media_player.MediaPlayerEntityFeature.VOLUME_SET
      required: false
      selector:
        number:
          min: 0
          max: 1
          step: 0.01
```

**Comment by aodtcr** [comment]
Hi all, just arrived at this discussion after looking into changing the default volume_step for a new media player (AVR). The default for the AVR is + or - 0.5 dB (remote or volume knob) while changing the volume with the buttons in HA results in + or - 10 dB. (10% of 100 step range). This is very undesirable behavior. 

I'd love to see a way to either have this smaller by default (e.g. 1%) or a per device (integration) or global setting for personal preference and tweaking.

**Comment by godehardt** [comment]
How can I use this step feature in an automation?  Can I use it for the action itself or is it configured for the entity?  If so, how?
Using HA 2025.2.4
Thanks in advance

**Reply by MartinHjelmare** [reply]
This is an ongoing architecture discussion. It's not documenting an existing feature.

**Comment by andrewsayre** [comment]
@glenrobertson I think this is worth implementing. Some integrations have added config options to set an integration default volume step because the default base implementation of 10% is too much. For some use cases, that wouldn't have been necessary if the services had this parameter.

As another datapoint, HEOS also has native support for volume up/down with a specified step (default is 5%). So that's another integration that would benefit from this feature.

Per @emontnemery comment, this proposal appears to be approved. Do you want to work on the implementation? If not, I'm happy to move it forward.

**Reply by glenrobertson** [reply]
Ok sounds good. I can take a look at it! Thanks for the background.

**Comment by glenrobertson** [comment]
@andrewsayre @aodtcr @emontnemery @gjohansson-ST  I have some changes in https://github.com/home-assistant/core/pull/140136 that I have confirmed work locally when testing with the `anthemav` component.
Please let me know what you think! I can add tests/docs, etc if the implementation is in the right direction.

**Comment by glenrobertson** [comment]
@emontnemery @andrewsayre @gjohansson-ST @godehardt @aodtcr do you all agree that it would be better to pass volume_step to  the volume_set command, as opposed to the volume_up/volume_down commands?

The benefit would be that the caller would have to only have to pass the step amount once, as opposed to passing it every time they want to change the volume up or down which seems less ideal.

This was the direction I took in the PR: https://github.com/home-assistant/core/pull/140136

**Reply by emontnemery** [reply]
Do you mean the benefit is that after passing a `volume_step` to the `volume_set` service, that step is remembered and respected by future calls to `volume_up` and `volume_down`?

I don't think this is a good idea, it means automations targeting the same media player will interfere with each other if they set different step sizes.

**Reply by arturpragacz** [reply]
I agree, that is definitely not a good idea.

But also I'm not sure whether adding `volume_step` to `volume_up` and `volume_down` is the best idea either.

What I like the most is the alternative proposed [here](https://github.com/home-assistant/architecture/discussions/891#discussioncomment-7680048)

> Hence, it would be better to add a `volume_step` parameter to `media_player.volume_set` instead, where `volume_step` would be a positive or negative float. This would be similar to `brightness_step` for `light.turn_on`.

Any reason we didn't decide to go with that one?

**Reply by emontnemery** [reply]
We already have the `volume_up` and `volume_down` services. Adding a step size to `volume_set` but keeping the `volume_up` and `volume_down` step size undefined seems illogical.

I think we should either implement the approved proposal, which is to add an optional `volume_step` to the `volume_up` and `volume_down` services, or extend it to add an optional `volume_step` to the `volume_up`, `volume_down` and `set_volume` services.

In no case should a the step size of previous service calls affect the behavior of future service calls.

**Reply by glenrobertson** [reply]
Got it. I closed the PR I had and don't really want to spend any more time on this. Anyone can feel free to take it on.
